test/e2e/storage/non_graceful_node_shutdown.go

```diff
@@ -0,0 +1,186 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package storage
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/onsi/ginkgo/v2"
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2enode "k8s.io/kubernetes/test/e2e/framework/node"
+	e2epod "k8s.io/kubernetes/test/e2e/framework/pod"
+	e2epv "k8s.io/kubernetes/test/e2e/framework/pv"
+	e2eskipper "k8s.io/kubernetes/test/e2e/framework/skipper"
+	"k8s.io/kubernetes/test/e2e/storage/drivers"
+	storageframework "k8s.io/kubernetes/test/e2e/storage/framework"
+	"k8s.io/kubernetes/test/e2e/storage/utils"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+	admissionapi "k8s.io/pod-security-admission/api"
+)
+
+/*
+This test assumes the following:
+- The infra is GCP.
+- NodeOutOfServiceVolumeDetach feature is enabled.
+
+This test performs the following:
+- Deploys a gce-pd csi driver
+- Creates a gce-pd csi storage class
+- Creates a pvc using the created gce-pd storage class
+- Creates an app deployment with replica count 1 and uses the created pvc for volume
+- Shutdowns the kubelet of node on which the app pod is scheduled.
+  This shutdown is a non graceful shutdown as by default the grace period is 0 on Kubelet.
+- Adds `out-of-service` taint on the node which is shut down.
+- Verifies that pod gets immediately scheduled to a different node and gets into running and ready state.
+- Starts the kubelet back.
+- Removes the `out-of-service` taint from the node.
+*/
+
+var _ = utils.SIGDescribe("[Feature:NodeOutOfServiceVolumeDetach] [Disruptive] [LinuxOnly] NonGracefulNodeShutdown", func() {
+	var (
+		c  clientset.Interface
+		ns string
+	)
+	f := framework.NewDefaultFramework("non-graceful-shutdown")
+	f.NamespacePodSecurityEnforceLevel = admissionapi.LevelPrivileged
+
+	ginkgo.BeforeEach(func() {
+		c = f.ClientSet
+		ns = f.Namespace.Name
+	})
+
+	ginkgo.Describe("NonGracefulNodeShutdown", func() {
+		ginkgo.Context("pod that uses a persistent volume via gce pd driver", func() {
+			ginkgo.It("should get immediately rescheduled to a different node after non graceful node shutdown ", func() {
+				// Install gce pd csi driver
+				ginkgo.By("deploying csi gce-pd driver")
+				driver := drivers.InitGcePDCSIDriver()
```

Comment:
Do we need to only enable for gce environment?

Reply:
Just need a csi driver to create volume dynamically. It does not matter what CSI driver or infra it is as far as testing the feature is concerned.  I choose gce here as the driver installation framework is already present.

Reply:
Kubernetes tests run in a variety of environment and clearly `InitGcePDCSIDriver` step would likely fail in AWS environment for example. If the test was written in a way that did not initialized specific driver, then yes it won't have mattered, but as it is written test will fail in environments other than GCE and hence should be either skipped in those environments or written using generic driver.

Reply:
It makes sense. But I was under assumption that this e2e will only in GCE environment as I have added it to test grid in the following way: 
https://github.com/kubernetes/test-infra/pull/26947/files

But if identifying the infra provider and skipping the test if the infra is not GCE (e.g. in this case) is the recommended way -- sure, I will add it.

Reply:
Kubernetes testsuites are also ran by downstream Kubernetes distributions( such as Openshift, EKS, AKS and many others)  and hence jobs added to test-infra projects are not the only jobs running these tests.

Reply:
Sure. Make sense. I will add a skipper if the infra is not GCE.